File: solution-typescript/src/monitoring/endpoints.ts
/**
 * Monitoring endpoints for TypeScript Solution
 */

import { Request, Response } from 'express';
import { prometheusMetrics } from './metrics';
import { healthChecker } from './health';
import { logger } from '../utils/logger';

export class MonitoringEndpoints {
  private static instance: MonitoringEndpoints;

  private constructor() {}

  public static getInstance(): MonitoringEndpoints {
    if (!MonitoringEndpoints.instance) {
      MonitoringEndpoints.instance = new MonitoringEndpoints();
    }
    return MonitoringEndpoints.instance;
  }

  public async getMetrics(req: Request, res: Response): Promise<void> {
    try {
      const metrics = prometheusMetrics.getMetrics();
      
      res.set('Content-Type', 'text/plain; version=0.0.4; charset=utf-8');
      res.send(metrics);
      
      logger.debug('Metrics endpoint accessed', {
        userAgent: req.get('user-agent'),
        correlationId: (req as any).correlationId,
      });
    } catch (error) {
      logger.error('Failed to serve metrics', { error: error instanceof Error ? error.message : 'Unknown error' });
      res.status(500).json({
        error: 'Failed to serve metrics',
        message: error instanceof Error ? error.message : 'Unknown error',
      });
    }
  }

  public async getHealth(req: Request, res: Response): Promise<void> {
    try {
      const type = req.query.type as string || 'liveness';
      let healthSummary;

      switch (type) {
        case 'liveness':
          healthSummary = await healthChecker.checkLiveness();
          break;
        case 'readiness':
          healthSummary = await healthChecker.checkReadiness();
          break;
        case 'deep':
          healthSummary = await healthChecker.checkDeepHealth();
          break;
        default:
          res.status(400).json({
            error: 'Invalid health check type',
            validTypes: ['liveness', 'readiness', 'deep'],
          });
          return;
      }

      const statusCode = this.getHealthStatusCode(healthSummary.overallStatus);
      
      res.status(statusCode).json({
        status: healthSummary.overallStatus,
        timestamp: healthSummary.timestamp,
        checks: healthSummary.results,
        summary: {
          total: healthSummary.totalChecks,
          healthy: healthSummary.healthyChecks,
          degraded: healthSummary.degradedChecks,
          unhealthy: healthSummary.unhealthyChecks,
        },
      });

      logger.debug('Health check endpoint accessed', {
        type,
        status: healthSummary.overallStatus,
        correlationId: (req as any).correlationId,
      });
    } catch (error) {
      logger.error('Health check failed', { error: error instanceof Error ? error.message : 'Unknown error' });
      res.status(500).json({
        status: 'unhealthy',
        error: 'Health check failed',
        message: error instanceof Error ? error.message : 'Unknown error',
        timestamp: Date.now(),
      });
    }
  }

  public async getHealthLive(req: Request, res: Response): Promise<void> {
    try {
      const healthSummary = await healthChecker.checkLiveness();
      const statusCode = this.getHealthStatusCode(healthSummary.overallStatus);
      
      res.status(statusCode).json({
        status: healthSummary.overallStatus,
        timestamp: healthSummary.timestamp,
      });
    } catch (error) {
      logger.error('Liveness check failed', { error: error instanceof Error ? error.message : 'Unknown error' });
      res.status(500).json({
        status: 'unhealthy',
        error: 'Liveness check failed',
        timestamp: Date.now(),
      });
    }
  }

  public async getHealthReady(req: Request, res: Response): Promise<void> {
    try {
      const healthSummary = await healthChecker.checkReadiness();
      const statusCode = this.getHealthStatusCode(healthSummary.overallStatus);
      
      res.status(statusCode).json({
        status: healthSummary.overallStatus,
        timestamp: healthSummary.timestamp,
      });
    } catch (error) {
      logger.error('Readiness check failed', { error: error instanceof Error ? error.message : 'Unknown error' });
      res.status(500).json({
        status: 'unhealthy',
        error: 'Readiness check failed',
        timestamp: Date.now(),
      });
    }
  }

  public async getServerInfo(req: Request, res: Response): Promise<void> {
    try {
      const serverInfo = {
        name: 'OpenProject MCP Server - TypeScript',
        version: '1.0.0',
        architecture: 'TypeScript/Node.js',
        environment: process.env.NODE_ENV || 'development',
        uptime: process.uptime(),
        memory: process.memoryUsage(),
        nodeVersion: process.version,
        platform: process.platform,
        features: {
          monitoring: true,
          metrics: true,
          healthChecks: true,
          structuredLogging: true,
          correlationIds: true,
        },
        endpoints: {
          mcp: '/mcp',
          health: '/health',
          metrics: '/metrics',
          healthLive: '/health/live',
          healthReady: '/health/ready',
        },
      };

      res.json(serverInfo);

      logger.debug('Server info endpoint accessed', {
        correlationId: (req as any).correlationId,
      });
    } catch (error) {
      logger.error('Failed to get server info', { error: error instanceof Error ? error.message : 'Unknown error' });
      res.status(500).json({
        error: 'Failed to get server info',
        message: error instanceof Error ? error.message : 'Unknown error',
      });
    }
  }

  public async getNodejsMetrics(req: Request, res: Response): Promise<void> {
    try {
      const memoryUsage = process.memoryUsage();
      const cpuUsage = process.cpuUsage();
      const uptime = process.uptime();
      
      const metrics = {
        memory: {
          rss: memoryUsage.rss,
          heapTotal: memoryUsage.heapTotal,
          heapUsed: memoryUsage.heapUsed,
          external: memoryUsage.external,
          arrayBuffers: memoryUsage.arrayBuffers,
        },
        cpu: {
          user: cpuUsage.user,
          system: cpuUsage.system,
        },
        uptime,
        pid: process.pid,
        version: process.version,
        platform: process.platform,
        arch: process.arch,
      };

      res.json(metrics);

      logger.debug('Node.js metrics endpoint accessed', {
        correlationId: (req as any).correlationId,
      });
    } catch (error) {
      logger.error('Failed to get Node.js metrics', { error: error instanceof Error ? error.message : 'Unknown error' });
      res.status(500).json({
        error: 'Failed to get Node.js metrics',
        message: error instanceof Error ? error.message : 'Unknown error',
      });
    }
  }

  private getHealthStatusCode(status: string): number {
    switch (status) {
      case 'healthy':
        return 200;
      case 'degraded':
        return 200;
      case 'unhealthy':
        return 503;
      case 'unknown':
        return 500;
      default:
        return 500;
    }
  }
}

export const monitoringEndpoints = MonitoringEndpoints.getInstance();

File: solution-fastmcp/src/monitoring/endpoints.py
"""
Monitoring Endpoints for FastMCP Solution

This module provides HTTP endpoints for monitoring the FastMCP solution.
"""

import json
from typing import Dict, Any, Optional
from datetime import datetime

from aiohttp import web
from aiohttp.web import Request, Response

from .metrics import get_monitoring
from .health import get_health_checker, HealthStatus
from ..monitoring import logger


class MonitoringEndpoints:
    """Monitoring endpoints for FastMCP solution"""
    
    def __init__(self):
        self.monitoring = get_monitoring()
        self.health_checker = None
        self.logger = logger.bind(component="monitoring_endpoints")
    
    def set_health_checker(self, health_checker) -> None:
        """Set the health checker instance"""
        self.health_checker = health_checker
    
    async def liveness_check(self, request: Request) -> Response:
        """Liveness check endpoint"""
        try:
            async with self.monitoring.monitor_request("GET", "/health/live", dict(request.headers)):
                health_summary = await self.health_checker.check_liveness()
                
                status_code = 200 if health_summary.overall_status == HealthStatus.HEALTHY else 503
                
                response_data = {
                    "status": health_summary.overall_status.value,
                    "timestamp": health_summary.timestamp,
                    "service": "fastmcp-solution",
                    "checks": [
                        {
                            "name": result.name,
                            "status": result.status.value,
                            "duration_ms": result.duration_ms,
                            "message": result.message
                        }
                        for result in health_summary.results
                    ]
                }
                
                return web.json_response(
                    response_data,
                    status=status_code,
                    headers={"Content-Type": "application/json"}
                )
                
        except Exception as e:
            self.logger.error("Liveness check failed", error=str(e))
            return web.json_response(
                {"status": "error", "message": str(e)},
                status=500
            )
    
    async def readiness_check(self, request: Request) -> Response:
        """Readiness check endpoint"""
        try:
            async with self.monitoring.monitor_request("GET", "/health/ready", dict(request.headers)):
                health_summary = await self.health_checker.check_readiness()
                
                status_code = 200 if health_summary.overall_status == HealthStatus.HEALTHY else 503
                
                response_data = {
                    "status": health_summary.overall_status.value,
                    "timestamp": health_summary.timestamp,
                    "service": "fastmcp-solution",
                    "total_checks": health_summary.total_checks,
                    "healthy_checks": health_summary.healthy_checks,
                    "degraded_checks": health_summary.degraded_checks,
                    "unhealthy_checks": health_summary.unhealthy_checks,
                    "checks": [
                        {
                            "name": result.name,
                            "status": result.status.value,
                            "duration_ms": result.duration_ms,
                            "message": result.message,
                            "details": result.details
                        }
                        for result in health_summary.results
                    ]
                }
                
                return web.json_response(
                    response_data,
                    status=status_code,
                    headers={"Content-Type": "application/json"}
                )
                
        except Exception as e:
            self.logger.error("Readiness check failed", error=str(e))
            return web.json_response(
                {"status": "error", "message": str(e)},
                status=500
            )
    
    async def deep_health_check(self, request: Request) -> Response:
        """Deep health check endpoint"""
        try:
            async with self.monitoring.monitor_request("GET", "/health/deep", dict(request.headers)):
                health_summary = await self.health_checker.check_deep_health()
                
                status_code = 200 if health_summary.overall_status == HealthStatus.HEALTHY else 503
                
                response_data = {
                    "status": health_summary.overall_status.value,
                    "timestamp": health_summary.timestamp,
                    "service": "fastmcp-solution",
                    "total_checks": health_summary.total_checks,
                    "healthy_checks": health_summary.healthy_checks,
                    "degraded_checks": health_summary.degraded_checks,
                    "unhealthy_checks": health_summary.unhealthy_checks,
                    "checks": [
                        {
                            "name": result.name,
                            "status": result.status.value,
                            "duration_ms": result.duration_ms,
                            "message": result.message,
                            "details": result.details
                        }
                        for result in health_summary.results
                    ]
                }
                
                return web.json_response(
                    response_data,
                    status=status_code,
                    headers={"Content-Type": "application/json"}
                )
                
        except Exception as e:
            self.logger.error("Deep health check failed", error=str(e))
            return web.json_response(
                {"status": "error", "message": str(e)},
                status=500
            )
    
    async def metrics(self, request: Request) -> Response:
        """Prometheus metrics endpoint"""
        try:
            async with self.monitoring.monitor_request("GET", "/metrics", dict(request.headers)):
                metrics_data = self.monitoring.metrics.get_metrics()
                
                return Response(
                    text=metrics_data,
                    content_type="text/plain; version=0.0.4; charset=utf-8"
                )
                
        except Exception as e:
            self.logger.error("Metrics endpoint failed", error=str(e))
            return web.json_response(
                {"status": "error", "message": str(e)},
                status=500
            )
    
    async def info(self, request: Request) -> Response:
        """Service information endpoint"""
        try:
            async with self.monitoring.monitor_request("GET", "/info", dict(request.headers)):
                response_data = {
                    "service": "OpenProject FastMCP Server",
                    "version": "1.0.0",
                    "architecture": "fastmcp-native",
                    "python_version": "3.11",
                    "fastmcp_version": "2.10.6",
                    "mcp_version": "1.12.1",
                    "endpoints": {
                        "mcp": "/mcp",
                        "health": {
                            "live": "/health/live",
                            "ready": "/health/ready",
                            "deep": "/health/deep"
                        },
                        "metrics": "/metrics",
                        "info": "/info"
                    },
                    "features": {
                        "prometheus_metrics": True,
                        "structured_logging": True,
                        "health_checks": True,
                        "correlation_ids": True,
                        "sse_support": True
                    },
                    "documentation": "https://github.com/modelcontextprotocol/python-sdk",
                    "timestamp": datetime.now().isoformat()
                }
                
                return web.json_response(
                    response_data,
                    status=200,
                    headers={"Content-Type": "application/json"}
                )
                
        except Exception as e:
            self.logger.error("Info endpoint failed", error=str(e))
            return web.json_response(
                {"status": "error", "message": str(e)},
                status=500
            )
    
    async def ping(self, request: Request) -> Response:
        """Ping endpoint for basic connectivity test"""
        try:
            async with self.monitoring.monitor_request("GET", "/ping", dict(request.headers)):
                response_data = {
                    "message": "pong",
                    "service": "fastmcp-solution",
                    "timestamp": datetime.now().isoformat()
                }
                
                return web.json_response(
                    response_data,
                    status=200,
                    headers={"Content-Type": "application/json"}
                )
                
        except Exception as e:
            self.logger.error("Ping endpoint failed", error=str(e))
            return web.json_response(
                {"status": "error", "message": str(e)},
                status=500
            )


def create_monitoring_endpoints() -> MonitoringEndpoints:
    """Create monitoring endpoints instance"""
    return MonitoringEndpoints()
